File: app/src/main/java/com/hogl/eregister/connect/ChatActivity.java
package com.hogl.eregister.connect;

import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;

import android.Manifest;
import android.app.Activity;
import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.net.wifi.p2p.WifiP2pGroup;
import android.net.wifi.p2p.WifiP2pManager;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.hogl.eregister.R;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import javax.crypto.SecretKey;

public class ChatActivity extends AppCompatActivity {

    TextView statusText, protocolText;
    RecyclerView messagesView;
    Button sendButton;
    ServerClient serverClient;
    LinearLayout messageLayout;
    Button disconnectButton;
    String tag;


    Boolean isHost;
    String hostAddress;

    Activity thisAct;
    Context thisContext;

    String deviceName;
    ArrayList<String> authStrings;
    Integer authStep;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);
        Intent intent = getIntent();
        isHost = intent.getBooleanExtra("isHost", false);
        hostAddress = intent.getStringExtra("hostAddress");
        thisAct = this;
        thisContext = getApplicationContext();

        initComponents();
        setUpServer();
        if (isHost) tag = "CHAT-HOST";
        else tag = "CHAT-CLIENT";


    }

    public interface finishedInterface {
        String completed(SecretKey k);
    }

    public finishedInterface myInterface = (k) -> {

        Log.d("Auth", "AUTH SERVICE DONE!");


        serverClient.setMessagesChangedListener(new ServerClient.messagesChangedListener() {
//            @Override
//            public void onMessagesChangedListener() {
//                runOnUiThread(new Runnable() {
//                    @Override
//                    public void run() {
//                        adapter.notifyDataSetChanged();
//                        messagesView.smoothScrollToPosition(adapter.getItemCount());
//                    }
//                });
//            }
        });
        serverClient.setSecured(true, null);
        sendButton.setEnabled(true);
        protocolText.setText("SECURED");
        protocolText.setBackgroundColor(getResources().getColor(android.R.color.holo_green_light));
        return "Done";
    };


    private void setUpServer() {
        //Phone is host
        if (isHost) {
            statusText.setText("Host");
            serverClient = new ServerClient(authStrings);
        }
        //Phone is client
        else {
            statusText.setText("Client");
            serverClient = new ServerClient(hostAddress, authStrings);
        }


        serverClient.start();

        disconnectButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                disconnect();
            }
        });

        sendButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ExecutorService executor = Executors.newSingleThreadExecutor();
                executor.execute(new Runnable() {
                    @Override
                    public void run() {
                        if (true) {
//                            String escapedMessage = messageToSend.replace("\'", "\\'");

                            try {

                                InputStream is = null;

                                File fi = new File("data/data/"+thisContext.getPackageName()+ "/databases/eRegister_database");
                                if(fi.exists()) {
                                    Log.e("&&&&&", "File exists");
                                }
                                is = new FileInputStream(fi);
                                ContentResolver cr = thisContext.getContentResolver();
//                                InputStream inputStream = cr.openInputStream(Uri.parse("data/data/"+thisContext.getPackageName()+ "/databases/eRegister_database"));
//                                is = getAssets().open("transfer_database");
                                byte[] fileBytes = new byte[is.available()];
                                is.read(fileBytes);
                                is.close();

                                int offset = 0;
                                int AttributeDataLen = 244;
//
//                                String encryptedMessage = protocolUtils.encrypt((String.valueOf(fileBytes.length).getBytes(StandardCharsets.UTF_8)), sharedKey);
                                Log.d(tag + "size of byte array : ", String.valueOf(fileBytes.length));
                                String testLength = fileBytes.length+"";
                                byte[] siz = (testLength).getBytes(StandardCharsets.UTF_8);
                                serverClient.write(siz);


                                while (offset < fileBytes.length) {
                                    int size = fileBytes.length - offset;
                                    if (size > AttributeDataLen) {
                                        size = AttributeDataLen;
                                    }
                                    byte[] data = new byte[size];
                                    System.arraycopy(fileBytes, offset, data, 0, size);
                                    offset += size;

//                                    encryptedMessage = protocolUtils.encrypt(data, sharedKey);
                                    Log.d(tag + "-SENT", data.toString());
                                    serverClient.write(data);


                                }
                            } catch (IOException e) {
                                e.printStackTrace();
                            }
                        }
                    }
                });

                //Hides the keyboard
//                InputMethodManager imm = (InputMethodManager) thisAct.getSystemService(Activity.INPUT_METHOD_SERVICE);
//                View v = thisAct.getCurrentFocus();
//                imm.hideSoftInputFromWindow(v.getWindowToken(), 0);

            }
        });


    }

    private void initComponents() {
        statusText = findViewById(R.id.status_text);
        protocolText = findViewById(R.id.protocol_text);
        messagesView = findViewById(R.id.messages_view);
        sendButton = findViewById(R.id.send_button);
        sendButton.setEnabled(true);
        messageLayout = findViewById(R.id.linearLayoutMessages);
        disconnectButton = findViewById(R.id.disconnect_button);

        messageLayout.setVisibility(View.VISIBLE);

        deviceName = Settings.Global.getString(getContentResolver(), "device_name");
        authStrings = new ArrayList<String>();
        authStep = 0;


    }

    public static void triggerRebirth(Context context) {
        Intent intent = new Intent(context, MainActivity.class);
        intent.addFlags(FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(intent);
        if (context instanceof Activity) {
            ((Activity) context).finish();
        }

        Runtime.getRuntime().exit(0);
    }


    private void disconnect() {
        Executors.newSingleThreadExecutor().execute(new Runnable() {
            @Override
            public void run() {

                WifiP2pManager manager = (WifiP2pManager) getSystemService(Context.WIFI_P2P_SERVICE);
                WifiP2pManager.Channel channel = manager.initialize(thisContext, getMainLooper(), null);


                if (channel != null) {

                    if (ActivityCompat.checkSelfPermission(thisContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                        //permission check
                    }
                    manager.requestGroupInfo(channel, new WifiP2pManager.GroupInfoListener() {
                        @Override
                        public void onGroupInfoAvailable(WifiP2pGroup group) {
                            if (group != null) {
                                manager.removeGroup(channel, new WifiP2pManager.ActionListener() {

                                    @Override
                                    public void onSuccess() {
                                        Log.d("p2p-disconnect", "removeGroup onSuccess -");
                                        triggerRebirth(thisContext);
                                    }

                                    @Override
                                    public void onFailure(int reason) {
                                        Log.d("p2p-disconnect", "removeGroup onFailure -" + reason);
                                        triggerRebirth(thisContext);
                                    }
                                });
                            } else {
                                triggerRebirth(thisContext);
                            }
                        }
                    });
                } else {
                    Log.d("DISCONNECT", "channel is null");
                }
            }
        });

    }
}

File: app/src/main/java/com/hogl/eregister/connect/ChatActivity.kt
package com.hogl.eregister.connect

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.net.wifi.p2p.WifiP2pManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.util.Log
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.Gson
import com.hogl.eregister.R
import com.hogl.eregister.activities.HomeActivity
import com.hogl.eregister.connect.ServerClient.messagesChangedListener
import com.hogl.eregister.data.InitApplication
import com.hogl.eregister.data.models.*
import org.json.JSONArray
import org.json.JSONObject
import java.io.*
import java.nio.charset.StandardCharsets
import java.util.concurrent.Executors
import javax.crypto.SecretKey
import kotlin.properties.Delegates

class ChatActivity : AppCompatActivity() {
    lateinit var statusText: TextView
    lateinit var protocolText: TextView
    lateinit var messagesView: RecyclerView
    lateinit var sendButton: Button
    lateinit var serverClient: ServerClient
    lateinit var messageLayout: LinearLayout
    lateinit var disconnectButton: Button
    lateinit var tag: String
    lateinit var handler: Handler
    var isHost by Delegates.notNull<Boolean>()
    var hostAddress: String = ""
    lateinit var thisAct: Activity
    lateinit var thisContext: Context
    lateinit var deviceName: String
    lateinit var authStrings: ArrayList<String>
    var authStep by Delegates.notNull<Int>()
    private lateinit var sharedPreferences: SharedPreferences

    private val movementViewModel: MovementViewModel by viewModels {
        MovementViewModelFactory((this.application as InitApplication).movementRepository)
    }
    private val guardViewModel: GuardViewModel by viewModels {
        GuardViewModelFactory((this.application as InitApplication).guardRepository)
    }
    private val visitorViewModel: VisitorViewModel by viewModels {
        VisitorViewModelFactory((this.application as InitApplication).visitorRepository)
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat)
        val intent = intent
        isHost = intent.getBooleanExtra("isHost", false)
        hostAddress = intent.getStringExtra("hostAddress").toString()
        thisAct = this
        thisContext = applicationContext
        handler = Handler(this.getMainLooper())
        initComponents()
        setUpServer()
        tag = if (isHost!!) "CHAT-HOST" else "CHAT-CLIENT"
    }

    interface finishedInterface {
        fun completed(k: SecretKey?): String?
    }

    var myInterface: finishedInterface = object : finishedInterface {
        override fun completed(k: SecretKey?): String? {
            Log.d("Auth", "AUTH SERVICE DONE!")
            serverClient!!.setMessagesChangedListener(object :
                messagesChangedListener { //            @Override
                //            public void onMessagesChangedListener() {
                //                runOnUiThread(new Runnable() {
                //                    @Override
                //                    public void run() {
                //                        adapter.notifyDataSetChanged();
                //                        messagesView.smoothScrollToPosition(adapter.getItemCount());
                //                    }
                //                });
                //            }
            })
            serverClient!!.setSecured(true, null)
            sendButton!!.isEnabled = true
            protocolText!!.text = "SECURED"
            protocolText!!.setBackgroundColor(resources.getColor(android.R.color.holo_green_light))
            return "Done"
        }
    }

    private fun setUpServer() {
        //Phone is host
        if (isHost!!) {
            statusText!!.text = "Host"
            serverClient = ServerClient(authStrings, thisContext)
        } else {
            statusText!!.text = "Client"
            serverClient = ServerClient(hostAddress, authStrings, thisContext)
        }
        serverClient!!.start()
        disconnectButton!!.setOnClickListener { disconnect() }

        sendButton!!.setOnClickListener {

            var database = JSONObject()
            visitorViewModel.allVisitors.observe(this) { visitors ->
                for (visitor in visitors) {
                    val visitorJson = Gson().toJson(visitor)
                    database.accumulate("visitors", JSONObject(visitor.toString()))
                }
            }
            movementViewModel.allMovements.observe(this) { movements ->
                for (movement in movements) {
                    val movementJson = Gson().toJson(movement)
                    database.accumulate("movements", JSONObject(movement.toString()))
                }
            }
            guardViewModel.allGuards.observe(this) { guards ->
                for (guard in guards) {
                    val guardJson = Gson().toJson(guard)
                    database.accumulate("guards", guardJson)
                }
            }
            if(database.length() > 0) {
                val jsonString = database.toString()
               this.saveJson(jsonString)
            }
            Thread.sleep(1000)
            val executor = Executors.newSingleThreadExecutor()

            executor.execute {
                if (true) {
                    try {
                        var `is`: InputStream? = null
                        val fi =
                            File("data/data/com.hogl.eregister/databases/database.json")
                        if (fi.exists()) {
                            Log.e("&&&&&", "File exists")
                        }
                        `is` = FileInputStream(fi)
                        val cr = thisContext!!.contentResolver
                        //                                InputStream inputStream = cr.openInputStream(Uri.parse("data/data/"+thisContext.getPackageName()+ "/databases/eRegister_database"));
                        //                                is = getAssets().open("transfer_database");
                        val fileBytes = ByteArray(`is`.available())
                        `is`.read(fileBytes)
                        `is`.close()
                        var offset = 0
                        val AttributeDataLen = 244
                        //
                        //                                String encryptedMessage = protocolUtils.encrypt((String.valueOf(fileBytes.length).getBytes(StandardCharsets.UTF_8)), sharedKey);
                        Log.d(
                            tag + "size of byte array : ",
                            fileBytes.size.toString()
                        )
                        val testLength = fileBytes.size.toString() + ""
                        val siz =
                            testLength.toByteArray(StandardCharsets.UTF_8)
                        serverClient!!.write(siz)
                        while (offset < fileBytes.size) {
                            var size = fileBytes.size - offset
                            if (size > AttributeDataLen) {
                                size = AttributeDataLen
                            }
                            val data = ByteArray(size)
                            System.arraycopy(fileBytes, offset, data, 0, size)
                            offset += size

                            //                                    encryptedMessage = protocolUtils.encrypt(data, sharedKey);
                            Log.d("$tag-SENT", data.toString())
                            serverClient!!.write(data)
                        }
                        handler!!.post {
                            Toast.makeText(
                                thisContext,
                                "Synchronize finished",
                                Toast.LENGTH_SHORT
                            ).show()
//                            val intent = Intent(thisAct, HomeActivity::class.java)
//                            startActivity(intent)
                        }
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                }


            }

            //Hides the keyboard
            //                InputMethodManager imm = (InputMethodManager) thisAct.getSystemService(Activity.INPUT_METHOD_SERVICE);
            //                View v = thisAct.getCurrentFocus();
            //                imm.hideSoftInputFromWindow(v.getWindowToken(), 0);
        }
    }

    private fun initComponents() {
        statusText = findViewById(R.id.status_text)
        protocolText = findViewById(R.id.protocol_text)
        sendButton = findViewById(R.id.send_button)
        disconnectButton = findViewById(R.id.disconnect_button)
        deviceName = Settings.Global.getString(contentResolver, "device_name")
        authStrings = ArrayList()
        authStep = 0
    }

    private fun saveJson(s: String) {
        val output:Writer

        val file = File("data/data/com.hogl.eregister/databases/database.json")
        if (!file.exists()) {
            file.createNewFile()
        }
        output=BufferedWriter(FileWriter(file))
        output.write(s)
        output.close()
    }

    private fun disconnect() {
        Executors.newSingleThreadExecutor().execute {
            val manager =
                getSystemService(WIFI_P2P_SERVICE) as WifiP2pManager
            val channel = manager.initialize(
                thisContext,
                mainLooper, null
            )
            if (channel != null) {
                if (ActivityCompat.checkSelfPermission(
                        thisContext!!,
                        Manifest.permission.ACCESS_FINE_LOCATION
                    ) != PackageManager.PERMISSION_GRANTED
                ) {
                    //permission check
                }
                manager.requestGroupInfo(
                    channel
                ) { group ->
                    if (group != null) {
                        manager.removeGroup(channel, object : WifiP2pManager.ActionListener {
                            override fun onSuccess() {
                                Log.d("p2p-disconnect", "removeGroup onSuccess -")
                                triggerRebirth(thisContext)
                            }

                            override fun onFailure(reason: Int) {
                                Log.d(
                                    "p2p-disconnect",
                                    "removeGroup onFailure -$reason"
                                )
                                triggerRebirth(thisContext)
                            }
                        })
                    } else {
                        triggerRebirth(thisContext)
                    }
                }
            } else {
                Log.d("DISCONNECT", "channel is null")
            }
        }
    }

    companion object {
        fun triggerRebirth(context: Context?) {
            val intent = Intent(context, MainActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context!!.startActivity(intent)
            if (context is Activity) {
                context.finish()
            }
            Runtime.getRuntime().exit(0)
        }
    }
}